neurobooth_os/iout/lsl_streamer.py

```diff
@@ -1,163 +1,291 @@
-# -*- coding: utf-8 -*-
-"""
-Created on Tue Nov 24 15:41:42 2020
-
-@author: adona
-"""
-import time
 import logging
-from neurobooth_os import config
+import threading
 from neurobooth_os.iout import metadator as meta
-from neurobooth_os.iout.mbient import Mbient
-
-
-def start_lsl_threads(node_name, collection_id="mvp_030", win=None, conn=None):
-    """Initiate devices and LSL streams based on databased parameters.
-
-    Parameters
-    ----------
-    node_name : str
-        Name of the server where to start the lsl threads
-    collection_id : str, optional
-        Name of studies collection in the database, by default "mvp_025"
-    win : object, optional
-        Pycharm window, by default None
-    conn : object, optional
-        Connector to the database, by default None
-
-    Returns
-    -------
-    dict of streams
-        Contains the name of the device and device class
-    """
-
-    if conn is None:
-        print("getting conn")
-        conn = meta.get_conn(database)
-
-    logger = logging.getLogger('session')
-
-    # Get params from all tasks
-    kwarg_devs = meta._get_device_kwargs_by_task(collection_id, conn)
-    # Get all device params from session
-    kwarg_alldevs = {}
-    for dc in kwarg_devs.values():
-        kwarg_alldevs.update(dc)
-
-    streams = {}
-    if node_name == "acquisition":
-        from neurobooth_os.iout.microphone import MicStream
-        from neurobooth_os.iout.camera_intel import VidRec_Intel
-        from neurobooth_os.iout.flir_cam import VidRec_Flir
-        from neurobooth_os.iout.iphone import IPhone
-
-        for kdev, argsdev in kwarg_alldevs.items():
-            if "Intel" in kdev:
-                logger.debug(f'LSL Streamer Starting: {kdev}')
-                streams[kdev] = VidRec_Intel(**argsdev)
-            elif "Mbient" in kdev:
-                # Don't connect mbients from STM
-                if any([d in kdev for d in ["Mbient_LF", "Mbient_RF"]]):
+from typing import Any, Dict, List, Callable
+from concurrent.futures import ThreadPoolExecutor, as_completed, wait
+
+
+# --------------------------------------------------------------------------------
+# Wrappers for device setup procedures.
+# TODO: Handle device setup calls and imports in a more standardized/extensible fashion!!!
+# --------------------------------------------------------------------------------
+def start_eyelink_stream(win, **device_args):
```

Comment:
Would it be worthwhile at this point to generalize these device start functions such that they take a device as an argument, and always call a (possibly no-op) device.start() function before returning?  They might also have to implement some kind of (possibly no-op) before_start() function that could do things like check for prepare() in the case of mbients, although that could possibly be handled in start()..

Reply:
I 100% think these should be generalized! The goal with this partial refactor was to highlight/organize all of these inconsistencies to help inform the final solution.

I think the real solution is that all devices should derive from a common Device base class that has methods defining the device lifecycle (defaulting to no-op). These inconsistencies should then be implemented in the appropriate lifecycle calls.